docs/content/sips/0xx-component-dependencies.md

```diff
@@ -0,0 +1,108 @@
+Created: May 31, 2024
+
+# Background
+This feature enables the polyglot re-use of components, allowing developers to specify how to use components written in various languages as libraries to fulfill dependencies in their Spin components.
+
+For instance, consider writing a Spin component in Go that depends on the `aws:client/s3` interface (expressed as an import in the compiled Go component, e.g., `import aws:client/s3`). To satisfy this dependency, you can use a componentized S3 client written in Rust that exports an instance of the `aws:client/s3` interface. Using the proposed syntax in this SIP, Spin developers can instruct Spin on how to compose these components together to achieve this.
+
+# Proposal
+Developers can use the following syntax to specify that they want to use a component as a dependency, either from a local disk or via a remote component registry:
+
+## Using dependencies from disk (local)
+```toml
+[component."infra-dashboard".dependencies]
+# Components can satisfy dependencies by import name (e.g., "aws:client/s3")
+# using a component from disk. It is implicitly assumed that the
+# `my_aws_client.wasm` component exports an instance of the
+# `aws:client/s3` interface.
+"aws:client/s3" = { path = "my_aws_client.wasm" }
+
+# Optionally, explicitly specify the export name to use to satisfy the `aws:client/s3` interface.
+"aws:client/s3" = { path = "my_aws_client.wasm", export = "my-s3-client" }
+
+# Without an interface name, attempt to resolve every import of the "aws:client" package.
+"aws:client" = { path = "my_aws_client.wasm" }
+```
+
+> ⚠️ NOTE: Using the package and interface forms together (e.g. `aws:client` and `aws:client/s3` respectively) within the same dependencies section will result in an error to prevent resolution ambiguities.
+
+## Using dependencies from a registry (remote)
+```toml
+[component."infra-dashboard".dependencies]
+# Equivalent to { version = "1.0.0" , package = "aws:client"}
+"aws:client" = "1.0.0"
+
+# Use the `aws:client` component package to satisfy any number of "wasi:blobstore" imports ...
+"wasi:blobstore" = { registry = "my-registry.io", version = "0.1.0", package = "aws:client" }
+```
+
+## Dependency Isolation
+Statically composing the parent Spin component and its dependencies yields a component where the parent component and dependencies are no longer isolated from eachother. Each dependency will inherit the configuration of the parent component, which may be desireable in some library use-cases (e.g., allowing the `aws:client/s3` dependency to make outbound requests via `allowed_outbound_hosts`). However, making this the default behavior contradicts the explicit opt-in nature of capabilities for Spin (and the component model), so by default, component dependencies in Spin will not inherit configuration.
+
+To allow dependencies to inherit configuration from the parent Spin component, developers can set `dependencies_inherit_configuration = true`, as shown in the following example:
```

Comment:
What gets inherited here e.g. files?  Part of the original variables plan was to allow variables to be mapped into dependencies (specifically, with different names) - is that part of this?

Reply:
I'm struggling to recall, could you elaborate on the original variables plan? 

Would it be helpful to record in the SIP the set of configurations that are inherited, i.e.,:
- `allowed_outbound_hosts`
- `key_value_stores`
- `sqlite_databases`
- `variables`
- `ai_models`
- `files`
- `environment`

Reply:
The motivbation for variables was:

* I take dependencies on `alice:aws-s3` and `bob:aws-rds`.
* `alice:aws-s3` requires an AWS auth token in a variable named `aws-token`.
* `bob:aws-rds` requires an AWS auth token in a variable named `auth-for-services`. (This is why I stay away from Bob's libraries.)
* I would like to have one configuration variable that I can feed through to both - e.g. have a `aws-auth` application variable, and the Alice dep has `aws-token = "{{ aws-auth }}"` and the Bob dep has `auth-for-services = "{{ aws-auth }}"`.

Does that help answer the question?  If the answer is "you need to define both variables on the parent component and inherit" then that's fine, just seeking to clarify.

Yeah listing what gets inherited and what doesn't would be super useful!  Thanks!

Reply:
Ah yeah, thanks for the refresher! Yeah that would require including both variables on the parent component and inherit within this proposal.

Will update!